eclipse/src/saros/filesystem/EclipseFileImplV2.java

```diff
@@ -0,0 +1,149 @@
+package saros.filesystem;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URI;
+import java.nio.charset.Charset;
+import org.apache.log4j.Logger;
+import org.eclipse.core.filesystem.EFS;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.core.runtime.NullProgressMonitor;
+import org.eclipse.core.runtime.OperationCanceledException;
+
+/** Eclipse implementation of the Saros file interface. */
+// TODO rename to EclipseFile
+public class EclipseFileImplV2 extends EclipseResourceImplV2 implements IFile {
+  private static final Logger log = Logger.getLogger(EclipseFileImplV2.class);
+
+  /** @see EclipseResourceImplV2#EclipseResourceImplV2(EclipseReferencePointImpl, IPath) */
+  EclipseFileImplV2(EclipseReferencePointImpl referencePoint, IPath relativePath) {
+    super(referencePoint, relativePath);
+  }
+
+  @Override
+  public String getCharset() throws IOException {
+    try {
+      return getDelegate().getCharset();
+
+    } catch (CoreException e) {
+      throw new IOException(e);
+    }
+  }
+
+  @Override
+  public void setCharset(String charset) throws IOException {
+    if (charset == null) {
+      return;
+    }
+
+    // Check whether charset is valid and supported
+    Charset.forName(charset);
+
+    try {
+      updateFileEncoding(charset);
+
+    } catch (CoreException | OperationCanceledException e) {
+      throw new IOException(e);
+    }
+  }
+
+  /**
+   * Updates encoding of a file. An effort is made to use the inherited encoding if available.
+   *
+   * <p>Does nothing if the file does not exist.
+   *
+   * @param encoding the encoding that should be used
+   * @throws CoreException if setting the encoding failed
+   * @throws OperationCanceledException if the setting of the charset was canceled
+   * @see org.eclipse.core.resources.IFile#setCharset(String, IProgressMonitor)
+   */
+  private void updateFileEncoding(final String encoding)
+      throws CoreException, OperationCanceledException {
+
+    final org.eclipse.core.resources.IFile file = getDelegate();
+
+    if (!file.exists()) return;
+
+    String projectEncoding = file.getProject().getDefaultCharset();
```

Comment:
This was copied from the [previous Eclipse file implementation](https://github.com/saros-project/saros/blob/master/eclipse/src/saros/filesystem/EclipseFileImpl.java#L58). Originally, the logic was already present in the file activity handler and I just moved it to the file implementation when introducing `IFile.setCharset(...)` (see #913; added to `EclipseFileImpl` in https://github.com/saros-project/saros/pull/913/commits/af00eb1525a36f5c008f3c0da2048872b20e3993, removed the old implementation in `FileActivityConsumer` in https://github.com/saros-project/saros/pull/913/commits/ec53ad253a69e05782dcfbefabb45b05270b8023).

I only kept this logic around to avoid the case you already mentions where Saros causes the encoding settings file to grow in size quite severely as an entry is added for every shared file.

What should this logic do instead? Always just compare to the charset of the direct parent container?

Also, how does eclipse handle changing the charset? Is the content automatically updated or is it just re-read with the new encoding? As an example, if I set the encoding of a file to inherit the parent container encoding as it currently matches and then change the parent container encoding, would the child file still be readable?

Reply:
I have not tested it but I guess file.getCharset() should return the correct charset (either from the nearest parent upwards to the project root), So if it differs you should not need to check the project encoding. To answer you whole question, just play around with the stuff, My big guess is, if you change the encoding of a folder while an affected file is open in an editor nothing happens. As I said, you just change meta data not the binary content. The binary content will never change. I can only guess if you change the encoding while the file is opened in an editor after a save the binary representation is correctly written to the underlaying storage.

Reply:
Done. Adjusted to logic to check the default encoding of the direct parent container instead.